schemachanger: add OFFLINE as intermediate state when dropping descriptors

postamar opened the pull request:
The addition of a revertible transition makes it possible to combine
descriptor drops with other schema changes in a way that preserves their
revertibility while immediately making the dropped descriptor no longer
public.

Release note (sql change): DROP statements performed by the declarative
schema changer (which is the case by default) now transition the
descriptor states to OFFLINE in the initial schema change transaction,
before subsequently transitioning them to DROP in a subsequent
transaction executed by the schema change job. The behaviour of
changefeeds on dropped tables will be correspondingly affected.
Previously they'd throw an error about the table being dropped.
They may now do either that or throw an error about the table being
taken offline. Furthermore, it's possible for a concurrent backup
to see the table as OFFLINE before it reaches DROP and this may cause
the offline table to be included in the data. We don't expect the impact
of these changes to be significant, but they may nonetheless be noticeable.

Commit:
scexec: add MarkDescriptorAsPublic and MarkDescriptorAsOffline ops

Release note: None

Commit:
opgen: add OFFLINE as intermediate state when dropping descriptors

The addition of a revertible transition makes it possible to combine
descriptor drops with other schema changes in a way that preserves their
revertibility while immediately making the dropped descriptor no longer
public.

Release note (sql change): DROP statements performed by the declarative
schema changer (which is the case by default) now transition the
descriptor states to OFFLINE in the initial schema change transaction,
before subsequently transitioning them to DROP in a subsequent
transaction executed by the schema change job. The behaviour of
changefeeds on dropped tables will be correspondingly affected.
Previously they'd throw an error about the table being dropped.
They may now do either that or throw an error about the table being
taken offline. Furthermore, it's possible for a concurrent backup
to see the table as OFFLINE before it reaches DROP and this may cause
the offline table to be included in the data. We don't expect the impact
of these changes to be significant, but they may nonetheless be noticeable.

Commit:
descs: fix type hydration bug

This patch fixes an instance where we ignored offline descriptors by
mistake.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83915)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Well this worked out shockingly well. We should make sure that backups don't get sad about it. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83915#-:-N6KReIv1mPP13ks6bsz:b-3e4mou)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
backupresolver: fix descriptor filtering in targeting

This commit addresses two issues:
1. Until now, type descriptors were never properly filtered. It was
   therefore possible to include some with a DROP state in a backup.
2. OFFLINE schema descriptors were never included in targeted backups.

These issues were of little importance until now: we want to
use the OFFLINE state in the declarative schema changer.

Release note: None

Commit:
rewrite: fix schema changer state rewrite bugs

These bugs became apparent while introducing the OFFLINE state in the
declarative schema changer DROP path.

Release note: None

postamar commented:
There were, in fact, some wonky non-trivial bugs related to backups not caring enough about non-public type descriptors and the offline state in general.

postamar commented:
> One issue is that backups do want to backup OFFLINE descriptors, at least, I think they do.

I think so too, which is why I relaxed the filtering in the backup targeting to include more OFFLINE descriptor than it previously was. Previous behaviour was also slightly erroneous irrespective of current changes: an OFFLINE table would be included in a BACKUP DATABASE but not the parent OFFLINE schema.

I'll make sure this PR gets reviewed by Bulk-IO as these changes also have an impact on changefeeds, which now can error complaining about a table being offline instead of dropped.

ajwerner commented:
>an OFFLINE table would be included in a BACKUP DATABASE but not the parent OFFLINE schema.

Ah, I suspect we've never had an `OFFLINE` schema before

postamar commented:
Indeed. Also I don't think anyone could have noticed anything before the `public` schema got an explicit descriptor.

Commit:
sctest: raise timeout threshold for Backup test

The default timeout was sometimes causing CI flakes.
The recent changes to DROP behaviour caused more cases to be run
in this test suite.

This change can (and should) be reverted once we're able to parallelize
test runs better in CI.

Release note: None

ajwerner commented on pkg/sql/catalog/descs/descriptor.go:
```diff
@@ -186,7 +186,7 @@ func (tc *Collection) getDescriptorsByID(
 		// hydrated where and test the API boundary accordingly.
 		if table, isTable := desc.(catalog.TableDescriptor); isTable {
-			desc, err = tc.hydrateTypesInTableDesc(ctx, txn, table)
+			desc, err = tc.hydrateTypesInTableDescWitOptions(ctx, txn, table, flags.IncludeOffline, flags.AvoidLeased)
```
I fixed this typo on master, so this needs a rebase, otherwise it'll skew and to adopt the fix, otherwise it'll skew

ajwerner reviewed (approved):
LGTM -- this proved to be important for some changes I needed to make for DROP COLUMN. I'd prefer we merge it sooner rather than later.

postamar commented:
In a few hours I'll push a commit which addresses https://github.com/cockroachdb/cockroach/issues/84137 on top of this branch (and rebase).

postamar commented:
> In a few hours I'll push a commit which addresses https://github.com/cockroachdb/cockroach/issues/84137 on top of this branch (and rebase).

The opposite is true, I actually need to rebase this branch on top of another. This should not hinder any reviews.

postamar commented:
bors r+
